design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,599 @@
+  can be used as an `AwsEndpoint`.
+- **AWS Endpoint**: An endpoint with explicit signing configuration applied. AWS Endpoints need to contain region &
+  service metadata to control signing.
+- **Middleware**: A transformation applied to a request, prior to request dispatch
+- **Endpoint Parameters**: A code-generated structure for each service which contains service-specific (and general)
+  endpoint parameters.
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+### Overview
+
+SDKs will generate a new, public, `endpoint` module. The module will contain a [`Params`](#endpoint-params) structure
+and
+a [`DefaultResolver`](#the-default-endpoint-resolver). Supporting these modules, a private `endpoints_impl` module will
+be generated.
+> **Why generate two modules**?
+>
+> Generating two separate modules, `endpoint` and `endpoint_impl` ensures that we don't have namespace collisions
+> between hand-written and generated
+> code.
+
+SDK middleware will be updated to use the new [`smithy_types::Endpoint`](#the-endpoint-struct). During request
+construction in `make_operation`, a [smithy endpoint](#the-endpoint-struct) will be inserted into the property bag. The
+endpoint middleware will be updated to extract the Smithy endpoint from the property bag and set the request endpoint &
+signing information accordingly (see: [Converting to AWS Endpoint](#converting-a-smithy-endpoint-to-an-aws-endpoint).
+
+The following flow chart traces the endpoints 2.0 influence on a request via the green boxes.
+
+```mermaid
+flowchart TD
+    globalConfig("SDK global configuration (e.g. region provider, UseFIPS, etc.)")
+
+    serviceConfig("Modeled, service specific configuration information (clientContextParams)")
+
+    operationConfig("Operation-specific configuration (S3 Bucket, accountId, etc.)")
+
+    getObject["S3::GetObject"]
+
+    params["Create endpoint parameters"]
+
+    evaluate["Evaluate ruleset"]
+
+    rules["Generated Endpoint Ruleset for S3"]
+
+    middleware["Apply endpoint & properties to request via endpoint middleware"]
+
+
+
+    style getObject fill:green,stroke:#333,stroke-width:4px
+    style params fill:green,stroke:#333,stroke-width:4px
+    style evaluate fill:green,stroke:#333,stroke-width:4px
+    style middleware fill:green,stroke:#333,stroke-width:4px
+
+    getObject ==> params
+    globalConfig ---> params
+    operationConfig --> params
+    serviceConfig ---> params
+
+    rules --> evaluate
+    params --> evaluate
+    evaluate --> middleware
+```
+
+### Overriding Endpoints
+
+In the general case, users will not be impacted by Endpoints 2.0 with one exception: today, users can provide a
+global endpoint provider that can override different services. There is a single `ResolveAwsEndpoint` trait that
```

Comment:
Where does this trait live?

Reply:
_currently_ it's in `aws_smithy_http` but it's possible it should end up in `aws_smithy_types`